spec.md

```diff
@@ -0,0 +1,954 @@
+| 15 DATALOSS |  |  |  |  |
+| 16 UNAUTHENTICATED |  |  |  |  |
+
+The status `message` MUST contain a human readable description of error, if the status `code` is not `OK`.
+This string MAY be surfaced by CMI-Client to end users.
+
+## Protocol
+
+### Connectivity
+
+* A CMI-Client SHALL communicate with a Plugin using gRPC to access the `Identity`, and `Machine` services.
+  * proto3 SHOULD be used with gRPC, as per the [official recommendations](http://www.grpc.io/docs/guides/#protocol-buffer-versions).
+  * All Plugins SHALL implement the REQUIRED Identity & Machine service RPCs.
+    Support for OPTIONAL RPCs is optional.
+* The CMI-Client SHALL provide the listen-address for the Plugin by way of the `CMI_ENDPOINT` environment variable.
+  Plugin components SHALL create, bind, and listen for RPCs on the specified listen address.
+  * Only UNIX Domain Sockets MAY be used as endpoints.
+    This will likely change in a future version of this specification to support non-UNIX platforms.
+* All supported RPC services MUST be available at the listen address of the Plugin.
+
+### Security
+
+* The CMI-Client operator and Plugin Supervisor SHOULD take steps to ensure that any and all communication between the CMI-Client and Plugin Service are secured according to best practices.
+* Communication between a CMI-Client and a Plugin SHALL be transported over UNIX Domain Sockets.
+  * gRPC is compatible with UNIX Domain Sockets; it is the responsibility of the CMI-Client operator and Plugin Supervisor to properly secure access to the Domain Socket using OS filesystem ACLs and/or other OS-specific security context tooling.
+  * VM Provider’s supplying stand-alone Plugin controller appliances, or other remote components that are incompatible with UNIX Domain Sockets MUST provide a software component that proxies communication between a UNIX Domain Socket and the remote component(s).
+    Proxy components transporting communication over IP networks SHALL be responsible for securing communications over such networks.
+* Both the CMI-Client and Plugin SHOULD avoid accidental leakage of sensitive information (such as redacting such information from log files).
+
+### Debugging
+
+* Debugging and tracing are supported by external, CMI-independent additions and extensions to gRPC APIs, such as [OpenTracing](https://github.com/grpc-ecosystem/grpc-opentracing).
+
+## Configuration and Operation
+
+### General Configuration
+
+* The `CMI_ENDPOINT` environment variable SHALL be supplied to the Plugin by the Plugin Supervisor.
+* An operator SHALL configure the CMI-Client to connect to the Plugin via the listen address identified by `CMI_ENDPOINT` variable.
+* With exception to sensitive data, Plugin configuration SHOULD be specified by environment variables, whenever possible, instead of by command line flags or bind-mounted/injected files.
+
+
+#### Plugin Bootstrap Example
```

Comment:
The endpoint examples all seem to be unix socket endpoints. Can we have URL endpoints in the samples?

Reply:
Yes, that makes sense. Shall try to update the same.

Reply:
Updated.